domino/datasets.py

```diff
@@ -0,0 +1,198 @@
+MAX_WORKERS = 10
+MAX_UPLOAD_ATTEMPTS = 10
+MB = 2 ** 20  # 2^20 bytes - 1 Megabyte
+SLEEP_TIME_IN_SEC = 3
+UPLOAD_READ_TIMEOUT_IN_SEC = 30
+
+
+@dataclass
+class UploadChunk:
+    """ Class for keeping track of a dataset upload chunk."""
+    absolute_path: str
+    chunk_number: int
+    dataset_id: str
+    file_name: str
+    file_size: int
+    upload_key: str
+    identifier: str
+    relative_path: str
+    target_chunk_size: int
+    total_chunks: int
+
+
+class Uploader:
+    def __init__(
+            self,
+            csrf_no_check_header: {str, str},
+            dataset_id: str,
+            local_path_to_file_or_directory: str,
+            log: Logger,
+            request_manager: _HttpRequestManager,
+            routes: _Routes,
+            target_relative_path: str,
+
+            file_upload_setting: str,
+            max_workers: int,
+            target_chunk_size: int,
+
+    ):
+        self.csrf_no_check_header = csrf_no_check_header
+        self.dataset_id = dataset_id
+        self.local_path_file_or_directory = local_path_to_file_or_directory
+        self.log = log
+        self.request_manager = request_manager
+        self.routes = routes
+        self.target_chunk_size = target_chunk_size or 8 * MB
+        self.file_upload_setting = file_upload_setting or FILE_UPLOAD_SETTING_DEFAULT
+        self.max_workers = max_workers or MAX_WORKERS
+        self.target_relative_path = target_relative_path
+        self.upload_key = None  # this will be set once the session is started
+
+    def __enter__(self):
+        # creating upload session
+        start_upload_body = {
+            "filePaths": [],
+            "fileCollisionSetting": self.file_upload_setting
+        }
+        start_upload_url = self.routes.datasets_start_upload(self.dataset_id)
+        self.upload_key = self.request_manager.post(start_upload_url, json=start_upload_body).json()
+        if not self.upload_key:
+            raise RuntimeError(f"upload key for {self.dataset_id} not found. Session could not start.")
+        # starting uploads
+        try:
+            paths = self._upload()
+            return paths
+        except Exception as e:
+            self.log.error("Upload failed. See error for details.")
+            raise e
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        # catching errors
+        if not self.upload_key:
+            raise RuntimeError(f"upload key for {self.dataset_id} not found. Could not end session.")
+        if exc_type != None:
+            self.log.error(f"Upload for dataset {self.dataset_id} and file {self.local_path_file_or_directory} failed, "
+                           f"attempting to cancel session. Please try again.")
+            self.log.error(f"Error type: {exc_val}. Error message: {exc_tb}.")
+            if not isinstance(exc_type, ValueError):
+                self._cancel_upload_session() # is it is a ValueError, canceling session would fail
```

Comment:
what happen if this raise an exception? afaik `__exit__` must return True or False https://docs.python.org/3/library/contextlib.html#contextlib.ContextDecorator

Reply:
it raises it! modifying below.